src/relations/openfga.py

```diff
@@ -0,0 +1,142 @@
+
+    @log_event_handler(logger)
+    def _on_openfga_store_created(self, event: OpenFGAStoreCreateEvent):
+        """Handle OpenFGA relation created event.
+
+        Args:
+            event: The event triggered when the relation is created.
+        """
+        if not event.store_id:
+            logger.info(f"{event.relation.name} revoked, no store id")
+            return
+
+        token = event.token
+        if event.token_secret_id:
+            secret = self.charm.model.get_secret(id=event.token_secret_id)
+            secret_content = secret.get_content()
+            token = secret_content["token"]
+
+        if self.charm.unit.is_leader():
+            self.charm._state.openfga = {
+                "store_id": event.store_id,
+                "token": token,
+                "address": event.address,
+                "port": event.port,
+                "scheme": event.scheme,
+                "auth_model_id": None,
+            }
+
+        self.charm._update(event)
+
+    @log_event_handler(logger)
+    def _on_openfga_relation_broken(self, event) -> None:
+        """Handle broken relations with OpenFGA.
+
+        Args:
+            event: The event triggered when the relation changed.
+        """
+        if not self.charm._state.is_ready():
+            event.defer()
+            return
+
+        if self.charm.unit.is_leader():
+            self.charm._state.openfga = None
+            self.charm._update(event)
+
+    @log_event_handler(logger)
+    def _on_create_authorization_model_action(self, event):
+        """Handle OpenFGA create authorization model action.
+
+        Args:
+            event: The event triggered when the action is performed.
+        """
+        model = event.params["model"]
+        if not model:
+            event.fail("authorization model not specified")
+            return
+        try:
+            model_json = json.loads(model)
+        except json.decoder.JSONDecodeError as error:
+            event.fail(f"error occurred: {error}")
+            return
+
+        if not self.charm._state.openfga:
+            event.fail("missing openfga relation")
+            return
+        openfga_store_id = self.charm._state.openfga["store_id"]
+        openfga_token = self.charm._state.openfga["token"]
+        openfga_address = self.charm._state.openfga["address"]
+        openfga_port = self.charm._state.openfga["port"]
+        openfga_scheme = self.charm._state.openfga["scheme"]
+        url = f"{openfga_scheme}://{openfga_address}:{openfga_port}/stores/{openfga_store_id}/authorization-models"
+        headers = {"Content-Type": "application/json"}
+        if openfga_token:
+            headers["Authorization"] = f"Bearer {openfga_token}"
+
+        # do the post request
+        logger.info(f"posting to {url}, with headers {headers}")
+        try:
+            response = requests.post(url, json=model_json, headers=headers, timeout=10)
+        except requests.HTTPError as error:
```

Comment:
Are we just interested in `HTTPError`? you might want to consider catching as well `RequestException` for other errors.